### ISSUE TITLE
Traffic shaping proposal

### ISSUE THREAD

**Issue by shackijj** [WICG/protected-auction-services-discussion#96]
## Terminology
[Seller Frontend Service](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_api.md#sellerfrontend-service) (SFS) is the front-end service of the system that runs in the [trusted execution environment](https://github.com/privacysandbox/fledge-docs/blob/main/trusted_services_overview.md#trusted-execution-environment) on a supported cloud platform. The service receives requests from the [seller's ad service](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_api.md#seller-ad-service) to initiate the Protected Audience auction flow. Then the service orchestrates requests (in parallel) to Buyers / DSPs participating in the auction for bidding.

[Buyer Frontend Service](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_api.md#buyerfrontend-service) (BFS) is the front-end service of the system that runs in the [trusted execution environment](https://github.com/privacysandbox/fledge-docs/blob/main/trusted_services_overview.md#trusted-execution-environment) on a supported cloud platform. This service receives requests to generate bids from a [SellerFrontEnd service](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_api.md#sellerfrontend-service). This service fetches real-time bidding signals that are required for bidding and calls the Bidding service.

[ROMA](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/roma_bring_your_own_binary.md) is a C++ library used for executing untrusted code in a secure, isolated environment. It’s currently used for BFS and SFS

## Problem Statement
Hosting B&A servers is a considerable incremental cost driver, primarily for the DSPs. The DSP BuyerFrontEnd, Bidding Service, and Key/Value Service servers will be called for every combination of a DSP-owned Interest group and the SSP used to communicate that impression opportunity to the DSP, which could lead to thousands of calls per impression opportunity, inflating the processing costs greatly. Obviously most of these calls would be either redundant or duplicative and thus most of the incremental cost will be wasteful. Expected HW optimization is around 50%

## High-level Design
![image](https://github.com/user-attachments/assets/1e5b89d5-4a8c-4eeb-8eb7-6f901eb4051c)
### Writing encoded logs in Seller Frontend Service
To build models it’s required to log parameters of incoming requests such as interest groups and context signals. Buyer interest groups should be made available for Seller’s UDF which is responsible for filtering. As not to allow tracking users, the logs are encrypted and keys for decrypting them are managed in the same way as keys for decoding auction data from browsers (there are coordinators who provide the keys.

Parameters for encrypted logging are provided by Sellers ML engineers. In the configuration of Seller Frontend Service, there may be a section that describes what fields are logged or alternatively there should be API in UDFs.
```
function shouldSendRequestToBuyer(auctionConfig, trafficShapingParams) {
  const isLearning = Math.random() < 0.05
  if (isLearning) {
    return {
      requestSent: true,
      isLearning: isLearning
    }
  } else {
    const p = runInference('./path-to-model', {
      buyer: trafficShapingParams.buyer,
      interestGroups: trafficShapingParams.interestGroups,
      country: auctionConfig.sellerSignals.country,
      publisher: auctionConfig.sellerSignals.publisher
    })
    return {
      requestSent: p > 0.42,
      isLearning: isLearning
    }
  }
}
function scoreAd(adMetadata, bid, auctionConfig, trustedScoringSignals, browserSignals,
  directFromSellerSignals, crossOriginTrustedSignals, trafficShapingParams) {
  if (trafficShapingParams.isLearning) {
    writeEncryptedLog({
      bid: bid,
      buyer: trafficShapingParams.buyer,
      interestGroups: trafficShapingParams.interestGroups,
      country: auctionConfig.sellerSignals.country,
      publisher: auctionConfig.sellerSignals.publisher
    })
  }
}
```

### Introduction of Seller Model Builder Service
It’s a service that will build inference models based on encrypted logs from Seller Frontend Service. It’s hosted in TEE and sandboxed in the same fashion as Seller or Buyer Frontend Service. Considering the amount of traffic Seller Model Builder Service should be scalable. Keys for decoding are managed in the same way as keys for decoding auction data from browsers (there are coordinators who provide the keys).

The code for building models are provided by Sellers ML engineers. It could be deployed to a cloud storage and used by Seller Model Builder Service. Below is an example of the code that is used for building models.
```
# Libraries
import torch as T
import torch.nn as nn
import numpy as np
# Specify the device
device = T.device("cpu")
# Setup configuration of the model
class NeuralNet(nn.Module):
    def __init__(self, x_data_dim):
        super(NeuralNet, self).__init__()
        self.hidden_layer1 = nn.Linear(x_data_dim, 2)
    def forward(self, x):
        return T.log_softmax(self.hidden_layer1(x), dim=1)
# Training dataset
class SellerDataset(T.utils.data.Dataset):
    def __init__(self, src_file):
        tmp_x = self._load_features_from_encrypted_logs(src_file)
        tmp_y = self._load_labels_from_encrypted_logs(src_file)
        self.x_data = T.tensor(features, dtype=T.float32).to(device)
        self.y_data = T.tensor(features, dtype=T.long).to(device)
    def _load_features_from_encrypted_logs(src_file):
        # API for reading encrypted logs is used in this function
        ... 
    def _load_labels_from_encrypted_logs(src_file):
        # API for reading encrypted logs is used in this function
        ...
# Data for training
training_dataset = SellerDataset("some_path_to_encrypted_logs")
# Initialize a model
x_data_dim = training_dataset.x_data.shape[1]
model = NeuralNet(x_data_dim)
# Some code for model training
...
# Save the model
model_scripted = T.jit.script(model)
model_scripted.save('path-to-model')
```

### Feedback loop between Seller Model Builder Service and Seller Frontend Service
To deliver encrypted logs and encrypted models there should be automation that processes that.

### Metrics
1. We suggest tracking the following metrics, most of which should be also available for big slices of traffic (e.g., country, publisher, interest group)
2. Seller Model Builder Service. Cross entropy (log-likelihood), ROC AUC, the success rate on train, validation, and test datasets 
3. Seller Model Builder Service. Number of tries and successes per prediction bucket (each bucket contains requests with a prediction score in some predefined range, 5 buckets are enough)
4. Seller frontend service. Number of tries and successes and average prediction score per prediction bucket for the exploration and exploitation groups

## Questions
- What technology may be used in Seller Model Builder Service for building inference models?
- Does the Chrome team see violations of user’s privacy given that logs and models are encrypted and used only inside the sandbox?
- What effort is required to build all required infrastructure?
- Is the current debug mode enough to debug the traffic-shaping models? 
- What should A/B tests look like in the described setup? 
- What is the Chrome team opinion on usability for Sellers?

**Comment by csharrison** [comment]
Hi @shackijj, thank you for posting this feature request! To begin, I was wondering if you could expand a bit on the goals of the proposal. From my reading, it seems the primary goal is to alleviate the performance burden on the DSP, is that right? I am wondering how this stacks up with other potential goals of traffic shaping like reducing auction latency or improving SSP costs, especially given that operating such a traffic shaping solution requires the SSP to do work. Can you help me better understand the primary (and secondary) motivations of this proposal? One reason I ask this is that I am wondering if our proposed [chaffing design](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_chaffing.md) is compatible with this design, as it may mean that pruned DSPs still get called by the SSP (though responding to chaff requests should be cheap).


Secondly, I was wondering if you could share how much traffic shaping you do today (or could do today) without this feature i.e. in the untrusted seller ad server before calling B&A? If I understand correctly, you could in theory train a model on contextual-only features with labels from the winning buyer.


Finally, I wanted to point out that there are privacy implications with the proposed design. We allow training a model based on cross-site user data, then expose inferences based on this data (both in the form of network traffic to DSPs, and also in the form of the winning DSP). Currently in B&A, these inferences leak the current user’s input to the system (e.g.  the eligible interest groups) and are protected with various techniques like chaffing and k-anon.The introduction of an ML model training loop expands this to potentially leak arbitrary other users’ previous input, which is a novel privacy risk. There is prior art in this space that I wanted to point out, which is our [private model training (PMT) proposal](https://github.com/WICG/turtledove/blob/main/PA_private_model_training.md). While the initial target of PMT is to enable DSP predictions at bid-time, the overall privacy risks are very similar to this proposal, as we want to build a system to allow DSPs to train a model (based on cross-site user data). Our plan we outlined to ensure this was privacy safe was to protect the model with differential privacy.

**Comment by shackijj** [comment]
> I was wondering if you could expand a bit on the goals of the proposal. From my reading, it seems the primary goal is to alleviate the performance burden on the DSP, is that right? I am wondering how this stacks up with other potential goals of traffic shaping like reducing auction latency or improving SSP costs, especially given that operating such a traffic shaping solution requires the SSP to do work. Can you help me better understand the primary (and secondary) motivations of this proposal?

The primary goal here is to reduce CPU costs on the DSP side, which could also help preserve network costs for the SSP, if applicable and SSPs might be interested in reducing CPU costs on the DSP side because it provides them with a competitive advantage. Furthermore, we anticipate that DSPs will begin implementing QPS limits relatively quickly. In this scenario, effective traffic filtration becomes critical, as random throttling is not an efficient solution.

> One reason I ask this is that I am wondering if our proposed [chaffing design](https://github.com/privacysandbox/protected-auction-services-docs/blob/main/bidding_auction_services_chaffing.md) is compatible with this design, as it may mean that pruned DSPs still get called by the SSP (though responding to chaff requests should be cheap).

The chaffing is compatible with traffic shaping while they operate independently. In other words, traffic shaping should only apply to real traffic and chaff requests shouldn’t be used for training models.

Traffic shaping may affect calculations of the chaff requests amount.

> Secondly, I was wondering if you could share how much traffic shaping you do today (or could do today) without this feature i.e. in the untrusted seller ad server before calling B&A? If I understand correctly, you could in theory train a model on contextual-only features with labels from the winning buyer.

Currently we are filtering out ~40% of requests using contextual features only. However, it looks reasonable to suppose that we can filter additional 30% out of remaining 60% of requests (i.e. one half).

> Finally, I wanted to point out that there are privacy implications with the proposed design. We allow training a model based on cross-site user data, then expose inferences based on this data (both in the form of network traffic to DSPs, and also in the form of the winning DSP). Currently in B&A, these inferences leak the current user’s input to the system (e.g. the eligible interest groups) and are protected with various techniques like chaffing and k-anon. The introduction of an ML model training loop expands this to potentially leak arbitrary other users’ previous input, which is a novel privacy risk.

Our goal here was precisely to validate the privacy concerns from your perspective. As we point out below, in this aspect we would like to be aligned with the PMT approach ([turtledove/PA_private_model_training.md at main · WICG/turtledove](https://github.com/WICG/turtledove/blob/main/PA_private_model_training.md#training-private-models-with-encryptedmodelingsignals))  and use differential privacy as the core concept. If you believe that such an approach guarantees user privacy, we are happy to reuse it. Apart from the PMT architecture, we are ready to use other solutions, for example any optimizer that respects differencial privacy

> There is prior art in this space that I wanted to point out, which is our [private model training (PMT) proposal](https://github.com/WICG/turtledove/blob/main/PA_private_model_training.md). While the initial target of PMT is to enable DSP predictions at bid-time, the overall privacy risks are very similar to this proposal, as we want to build a system to allow DSPs to train a model (based on cross-site user data). Our plan we outlined to ensure this was privacy safe was to protect the model with differential privacy.

Indeed, the PMT proposal looks very similar to our idea, thanks for pointing it out! It seems that at least [this part](https://github.com/WICG/turtledove/blob/main/PA_private_model_training.md#training-private-models-with-encryptedmodelingsignals)  can be reused for our purposes almost unchanged (including differencial privacy). However, some points important for our proposal are not covered by the PMT design. For instance, the PMT approach assumes that encrypted logs or aggregates are written only in case of wins, while we need the data for all the learning requests sent to buyer frontends. Namely, we depend on the implementation of writeEncryptedLog in the scoreAd function. If it is possible to make this or a similar function part of the PMT proposal, we are open to discussion. In any case we can make our proposal more consistent with the PMT design in terms of naming and payload format. Can we expect a release of the PMT architecture in the near future?
